httpcore/_synchronization.py

```diff
@@ -2,17 +2,44 @@
 from types import TracebackType
 from typing import Optional, Type
 
-import anyio
+import sniffio
 
 from ._exceptions import ExceptionMapping, PoolTimeout, map_exceptions
 
+# Our async synchronization primatives use either 'anyio' or 'trio' depending
+# on if they're running under asyncio or trio.
+#
+# We take care to only lazily import whichever of these two we need.
+
 
 class AsyncLock:
     def __init__(self) -> None:
-        self._lock = anyio.Lock()
+        self._backend = ""
+
+    def setup(self) -> None:
+        """
+        Detect if we're running under 'asyncio' or 'trio' and create
+        a lock with the correct implementation.
+        """
+        self._backend = sniffio.current_async_library()
+        if self._backend == "trio":
+            import trio
+
+            self._trio_lock = trio.Lock()
+        else:
+            import anyio
+
+            self._anyio_lock = anyio.Lock()
```

Comment:
Do we need to differentiate the name, tho?
```suggestion
            import trio

            self._lock = trio.Lock()
        else:
            import anyio

            self._lock = anyio.Lock()
```

Reply:
I think we'd get awkward type errors if we used the same attribute for both cases. (?)